DEV-GUIDELINES.md

```diff
@@ -0,0 +1,187 @@
+3. Config keys:
+    1. Use lower case words separated by dashes 
+        (e.g. "token-endpoint-uri", NOT "tokenEndpointUri")
+    2. May be nested in a tree structure (e.g. outbound-token.name, outbound-token.algorithm)
+    3. Each component should have useful defaults defined for each property it expects. Components should only fail when missing 
+         a configuration that cannot be defaulted
+
+Example: [io.helidon.security.oidc.common.OidcConfig](security/providers/oidc-common/src/main/java/io/helidon/security/oidc/common/OidcConfig.java)
+
+# Getters and Setters
+1. We do not use the verb, e.g. when a property "port" exists, the following methods are used:  
+    1. port(int newPort)
+    2. int port()
+2. Boolean depends on how well understood the method is
+    1. Default is without a verb (e.g. authenticate(boolean atn), boolean authenticate())
+    2. If this would be ambiguous, we can use verb to clear the meaning (e.g. isAuthenticated() or shouldAuthenticate())      
+
+Example: [io.helidon.security.oidc.common.OidcConfig](security/providers/oidc-common/src/main/java/io/helidon/security/oidc/common/OidcConfig.java) 
+
+# Fluent API
+1. We use fluent API where applicable
+    1. In builders (see builders section below)
+    2. When using control methods (such as Server server = Server.create().start())
+    
+# Builders
+1. We use builders to create instances that need any parameters for construction
+    1. **This implies that there are no public API classes that would use public constructors**
+    2. Allowed exceptions to this rule:
+        1. Integration APIs that follow rules of integrated solution, e.g. Jersey SecurityFeature
+2. Class or interface using a builder (let's call ours "FooBar" for the purpose of this document)
+    1. Must have:   
+        1. Hidden constructor (private or protected) - this is to allow us to switch to interface if needed
+        2. method _public static Builder builder()_             
+        3. all fields obtained from builder declared as final (immutable)
+    2. May have:
+        1. method _public static Builder builder(?)_ - builder with mandatory or very commonly used parameters
+            - there should be a very small number of "builder(*)" methods - up to two per class
+        1. Factory method _public static FooBar create()_ that is implemented as "return builder().build()"
+        2. Factory method _public static create(io.helidon.config.Config config)_ that is implemented as 
+            "return builder().config(config).build()"
+        2. Other factory methods that build specific (predefined) instances, such as "fail(String cause)", 
+            "success(Subject subject)" etc. - **these methods MUST use builder to build the instance internally**
+        3. An internal "Builder" class used to build it - it is allowed to have the builder as a top level class,
+            though the name must reflect the class it is building (e.g. FooBarBuilder) 
```

Comment:
That is written in the rule.
1) It should be an internal Builder class
2) It may be a top level public class, in which case it must be named differently, as it is NOT contained in another class.

Reply:
Reworded